cmd/clusterctl/client/alpha/rollout_rollbacker.go

```diff
@@ -0,0 +1,171 @@
+	patchHelper, err := patch.NewHelper(d, c)
+	if err != nil {
+		return err
+	}
+	// Copy template into the machinedeployment (excluding the hash)
+	revMSTemplate := *msForRevision.Spec.Template.DeepCopy()
+	delete(revMSTemplate.Labels, mdutil.DefaultMachineDeploymentUniqueLabelKey)
+
+	d.Spec.Template = revMSTemplate
+	return patchHelper.Patch(context.TODO(), d)
+}
+
+// findMachineDeploymentRevision finds the specific revision in the machine sets
+func findMachineDeploymentRevision(toRevision int64, allMSs []*clusterv1.MachineSet) (*clusterv1.MachineSet, error) {
+
+	var (
+		latestMachineSet   *clusterv1.MachineSet
+		latestRevision     = int64(-1)
+		previousMachineSet *clusterv1.MachineSet
+		previousRevision   = int64(-1)
+	)
+	for _, ms := range allMSs {
+		if v, err := mdutil.Revision(ms); err == nil {
+			if toRevision == 0 {
+				if latestRevision < v {
+					// newest one we've seen so far
+					previousRevision = latestRevision
+					previousMachineSet = latestMachineSet
+					latestRevision = v
+					latestMachineSet = ms
+				} else if previousRevision < v {
+					// second newest one we've seen so far
+					previousRevision = v
+					previousMachineSet = ms
+				}
+			} else if toRevision == v {
+				return ms, nil
+			}
+		}
+	}
+
+	if toRevision > 0 {
+		return nil, errors.Errorf("unable to find specified revision: %v", toRevision)
+	}
+
+	if previousMachineSet == nil {
+		return nil, errors.Errorf("no rollout history found")
+	}
+	return previousMachineSet, nil
+
+}
+
+// getMachineSetsForDeployment returns a list of MachineSets associated with a MachineDeployment.
+func getMachineSetsForDeployment(proxy cluster.Proxy, d *clusterv1.MachineDeployment) ([]*clusterv1.MachineSet, error) {
+	log := logf.Log
+	c, err := proxy.NewClient()
+	if err != nil {
+		return nil, err
+	}
+	// List all MachineSets to find those we own but that no longer match our selector.
+	machineSets := &clusterv1.MachineSetList{}
+	if err := c.List(context.TODO(), machineSets, client.InNamespace(d.Namespace)); err != nil {
+		return nil, err
+	}
+
+	filtered := make([]*clusterv1.MachineSet, 0, len(machineSets.Items))
+	for idx := range machineSets.Items {
+		ms := &machineSets.Items[idx]
+
+		selector, err := metav1.LabelSelectorAsSelector(&d.Spec.Selector)
+		if err != nil {
+			log.V(5).Info("Skipping MachineSet, failed to get label selector from spec selector", "machineset", ms.Name)
+			continue
+		}
+
+		// If a MachineDeployment with a nil or empty selector creeps in, it should match nothing, not everything.
+		if selector.Empty() {
+			log.V(5).Info("Skipping MachineSet as the selector is empty", "machineset", ms.Name)
+			continue
+		}
+
+		// Skip this MachineSet if selector does not matche
+		if !selector.Matches(labels.Set(ms.Labels)) {
+			log.V(5).Info("Skipping MachineSet, label mismatch", "machineset", ms.Name)
+			continue
+		}
+		// Skip this MachineSet if its controller ref is not pointing to this MachineDeployment
```

Comment:
This check should probably go as first...

Reply:
done